pkg/providers/gcp/gcp.go

```diff
@@ -229,3 +273,75 @@ func (g *GCP) instanceMetrics(
 	}
 	return metrics, nil
 }
+
+// Refresh fetches all the Instances
+// for a project and  nd stores metadata in order to help with
+// metric collections
+func (g *GCP) Refresh(ctx context.Context, project string) {
+	iter := g.instances.AggregatedList(
+		ctx,
+		&computepb.AggregatedListInstancesRequest{
+			Project: project,
+		},
+	)
+
+	for {
+		resp, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			klog.Errorf("error while processes GCE instances %s", err)
+			return
+		}
+
+		for _, instance := range resp.Value.Instances {
+			zone := getValueFromURL(instance.GetZone())
+			instanceID := strconv.FormatUint(instance.GetId(), 10)
+			name := instance.GetName()
+
+			if zone == "" {
+				continue
+			}
+
+			if instance.GetStatus() == "TERMINATED" {
+				// delete the entry from the cache
+				g.cache.Delete(cacheKey(zone, service, name))
+				continue
+			}
+
+			if instance.GetStatus() == "RUNNING" {
+				// TODO potentially we do not need a custom resource to cache, maybe
+				// just cache the instance
+				g.cache.Add(cacheKey(zone, service, name), resource{
```

Comment:
To be explicit, may be worth while add a debug log for the "error" and say `item already exists, not adding` or whatever